### PR TITLE
algo_arena_and_results

### PR DESCRIPTION
**Pull request by idraper**
There are two files in this PR.

run_arena.py takes algos as inputs and then runs every combination of games for a set batch_size.

get_results.py is the beginning of a replay analyzer and has the ability to display various data from replay files.

Let me know if you have any comments, suggestions, or required edits :).

### CODE REVIEW

**Comment by idraper** [comment]
Alright, I created a header for each of those files. I also included a copyright so people know they can do whatever they want with it and also a short description for each.

I also added `#!/usr/bin/env python` for Unix users, since I forgot to put that in before. I made some small changes to the instructions for each to make them slightly more clear and up-to-date.

File: scripts/contributions/get_results.py [hunk 1]
```diff
@@ -0,0 +1,737 @@
...
+		default=[],
+		help="specify what data you would like to be graphed - you must have matplotlib installed\n\nValid Options For Single Game:\n\t- health\n\t- bits\n\t- cores\n\t- cores_spent\n\t- bits_spent\n\t- cores_on_board\n\nValid Options For Multiple Games:\n\t- wins\n\n")
+	return vars(ap.parse_args())
+
+
+class Graph:
+	numRows, numCols = 0,0
+	fig, ax, arg = None, None, None
+	pos = (0,0)
+	emptyPlots = []
```
Comment (876584635678890) [inline scripts/contributions/get_results.py]: Use of camel case, which is not idiomatic.

File: scripts/contributions/get_results.py [hunk 2]
```diff
@@ -0,0 +1,737 @@
...
+
+	verbose_options = ['health', 'bits', 'cores', 'cores_spent', 'bits_spent', 'cores_on_board']
```
Comment (876584635678890) [inline scripts/contributions/get_results.py]: Inconsistent use of camel case and underscores.

**Comment by 876584635678890** [comment]
Really like the idea of it.  
I did some testing and it did not work for me because my replays come without names for some reason, which results in crashing.
My review is not mandatory at all, but it is an improvement I would make.

**Comment by idraper** [comment]
Thanks for helping! I totally was switching between c++ and python naming conventions without even noticing, so thanks for pointing that out. I've gone through and changed everything (I hope) to be python in nature.

When you say you don't have name, do you mean the replay file doesn't have a name or the algo doesn't? If it is the algo, do you have two algos that have the same name?

File: scripts/contributions/get_results.py [hunk 3]
```diff
@@ -0,0 +1,736 @@
...
+this from your within game (although there is not really a reason to, since it looks at
+all the data after the replay is completed).
+
+I plan on adding more to this file, specifically the ability to graph much more
+data and get more useful statistics.
+
+If you have any suggestions/questions let me know :) - @Isaac
+'''
+
+pltInstalled = False
```
Comment (876584635678890) [inline scripts/contributions/get_results.py]:
```suggestion
plt_installed = False
```

File: scripts/contributions/get_results.py [hunk 4]
```diff
@@ -0,0 +1,736 @@
...
+	import glob
+	import math
+	import argparse
+except ImportError as e:
+	sys.stderr.write("WARNING: Module not found, full error:\n\n")
+	sys.stderr.write(e)
+
+try:
+	import matplotlib.pyplot as plt
+	pltInstalled = True
```
Comment (876584635678890) [inline scripts/contributions/get_results.py]:
```suggestion
	plt_installed = True
```

File: scripts/contributions/get_results.py [hunk 5]
```diff
@@ -0,0 +1,736 @@
...
+except ImportError:
+	try:
+		usrIn = input('Matplotlib not found.\nWould you like this program to try and install matplotlib? (y/n) ')
```
Comment (876584635678890) [inline scripts/contributions/get_results.py]:
```suggestion
		usr_in = input('Matplotlib not found.\nWould you like this program to try and install matplotlib? (y/n) ')
```

File: scripts/contributions/get_results.py [hunk 6]
```diff
@@ -0,0 +1,736 @@
...
+		if usrIn.lower() == 'y' or usrIn.lower() == 'yes':
```
Comment (876584635678890) [inline scripts/contributions/get_results.py]:
```suggestion
		if usr_in.lower() == 'y' or usr_in.lower() == 'yes':
```

File: scripts/contributions/get_results.py [hunk 7]
```diff
@@ -0,0 +1,736 @@
...
+			import subprocess
+			subprocess.run(['python', '-m', 'pip', 'install', 'matplotlib'])
+
+			import matplotlib.pyplot as plt
+			pltInstalled = True
```
Comment (876584635678890) [inline scripts/contributions/get_results.py]:
```suggestion
			plt_installed = True
```

File: scripts/contributions/get_results.py [hunk 8]
```diff
@@ -0,0 +1,736 @@
...
+			sys.stderr.write("\n\n")
+	except:
+		pltInstalled = False
```
Comment (876584635678890) [inline scripts/contributions/get_results.py]:
```suggestion
		plt_installed = False
```

File: scripts/contributions/get_results.py [hunk 9]
```diff
@@ -0,0 +1,736 @@
...
+
+def main(args):
+	verbose_options, summary_options = get_graph_options(args['graph'])
+
+	fh = FileHandler()
+	fh.load_files(int(args['num']), args['all'], args['file']) #loads the files - all JSON reading is here
+
+	# check to see if matplotlib is installed
+	graphing_enabled = True if len(args['graph']) > 0 else False
+	if graphing_enabled and not pltInstalled:
```
Comment (876584635678890) [inline scripts/contributions/get_results.py]:
```suggestion
	if graphing_enabled and not plt_installed:
```

File: scripts/contributions/get_results.py [hunk 10]
```diff
@@ -0,0 +1,736 @@
...
+		num_sub_plots = arg.count(':') + 1
+		r = math.floor(math.sqrt(num_sub_plots))
+		c = math.ceil(num_sub_plots/r)
+
+		fig_size = [6.4, 4.8]
+		max_x, max_y = 15, 9
+		fig_size = [fig_size[0] * c, fig_size[1] * r]
+		if fig_size[0] > max_x: fig_size[0] = max_x
+		if fig_size[1] > max_y: fig_size[1] = max_y
+		plt.rcParams["figure.figsize"] = fig_size
```
Comment (876584635678890) [inline scripts/contributions/get_results.py]: Not sure about this one.

File: scripts/contributions/get_results.py [hunk 11]
```diff
@@ -0,0 +1,736 @@
...
+
+# Stores data from a single replay and creates the Algo classes
+class Replay:
+	def __init__(self, f_name, algos):
+		self.fname = f_name;
+		self.ref = None
+		self.turns = {}
+		self.valid_turns = []
+
+		self.loadData()				# handles loading all the data from file into python variables
```
Comment (876584635678890) [inline scripts/contributions/get_results.py]:
```suggestion
		self.load_data()				# handles loading all the data from file into python variables
```

File: scripts/contributions/get_results.py [hunk 12]
```diff
@@ -0,0 +1,736 @@
...
+	def __eq__(self, other):
+		return self.fname == other.fname
+	def __toString(self):
+		return self.fname
+	def __str__(self):
+		return self.__toString()
+	def __repr__(self):
+		return self.__toString()
+
+	def loadData(self):
```
Comment (876584635678890) [inline scripts/contributions/get_results.py]:
```suggestion
	def load_data(self):
```

File: scripts/contributions/get_results.py [hunk 13]
```diff
@@ -0,0 +1,736 @@
...
+		self.wins = 0
+		self.cores_on_board = {}
+		self.replays = {} 	# this effectively holds all raw json information
+
+	# NOTE: eq will return true when comparing to strings of the same name - this is intentional to be able to use: str in listOfAlgos syntax.
+	def __eq__(self, other):
+		if type(other) == str:
+			return self.name == other
+		return self.name == other.name
+	def __toString(self):
```
Comment (876584635678890) [inline scripts/contributions/get_results.py]:
```suggestion
	def __string(self):
```

File: scripts/contributions/get_results.py [hunk 14]
```diff
@@ -0,0 +1,736 @@
...
+		return self.name
+	def __str__(self):
+		return self.__toString()
```
Comment (876584635678890) [inline scripts/contributions/get_results.py]:
```suggestion
		return self.__string()
```

File: scripts/contributions/get_results.py [hunk 15]
```diff
@@ -0,0 +1,736 @@
...
+	def __repr__(self):
+		return self.__toString()
```
Comment (876584635678890) [inline scripts/contributions/get_results.py]:
```suggestion
		return self.__string()
```

**Comment by idraper** [comment]
Thanks for the input. I changed all the additional naming errors you noticed. I also fixed the `'name'` bug you were experiencing.

I'll look into the matplotlib error now.

File: scripts/contributions/get_results.py (hunk 10 above)
Comment (idraper) [reply scripts/contributions/get_results.py]: `rcParams` is specific to matplotlib and cannot be changed.

**Comment by 876584635678890** [comment]
I fixed it, needed to install `tkinter`.

**Comment by idraper** [comment]
@876584635678890  really? How does that occur? Looking at the matplotlib dependencies:
Requires-Python: >=3.5
Requires-Dist: numpy (>=1.10.0)
Requires-Dist: cycler (>=0.10)
Requires-Dist: kiwisolver (>=1.0.1)
Requires-Dist: pyparsing (!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1)
Requires-Dist: python-dateutil (>=2.1)

**Comment by 876584635678890** [comment]
https://stackoverflow.com/q/36327134/6509751

Also, plotting works fine, but only when using `-v` as well. Otherwise it will only plot total wins.

**Comment by idraper** [comment]
That is intended use. If you run more than one game it defaults to "summary" view, which is only information about the aggregate of matches. The -v flag forces it to graph more information. This is information you don't necessarily care about if you are running it on multiple games.

**Comment by skelly1324** [comment]
Fantastic stuff here, this is great.

**Pull request merged**